system_tests/espresso_test.go

```diff
@@ -0,0 +1,282 @@
+package arbtest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/jarcoal/httpmock"
+	"github.com/offchainlabs/nitro/arbos/espresso"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/validator/server_api"
+	"github.com/offchainlabs/nitro/validator/valnode"
+)
+
+var (
+	validationPort = 54320
+	broadcastPort  = 9642
+)
+
+func espresso_block_txs_generators(t *testing.T, l2Info *BlockchainTestInfo) map[int][][]byte {
+	return map[int][][]byte{
+		5:  onlyMalformedTxs(t),
+		15: userTxs(t, l2Info),
+		25: func(t *testing.T) [][]byte {
+			// Contains malformed txes, valid transactions and invalid transactions
+			r := [][]byte{}
+			r = append(r, onlyMalformedTxs(t)...)
+			r = append(r, userTxs(t, l2Info)...)
+			return r
+		}(t),
+	}
+}
+
+func onlyMalformedTxs(t *testing.T) [][]byte {
+	return [][]byte{
+		{1, 2, 3},
+		{1, 2, 3},
+		{4, 5, 6},
+		{1, 2, 4},
+	}
+}
+
+// Two valid transactions and 2 invalid transactions with invalid nonces
+func userTxs(t *testing.T, l2Info *BlockchainTestInfo) [][]byte {
+	tx1 := l2Info.PrepareTx("Faucet", "Owner", 3e7, big.NewInt(1e16), nil)
+	tx1Bin, err := json.Marshal(tx1)
+	if err != nil {
+		panic(err)
+	}
+	tx2 := l2Info.PrepareTx("Owner", "Faucet", 3e7, big.NewInt(1e16), nil)
+	tx2Bin, err := json.Marshal(tx2)
+	if err != nil {
+		panic(err)
+	}
+	// 2 valid transactions here
+	return [][]byte{
+		tx1Bin,
+		tx1Bin,
+		tx2Bin,
+		tx2Bin,
+	}
+}
+
+func createMockHotShot(ctx context.Context, t *testing.T, l2Info *BlockchainTestInfo) (func(), int) {
+	httpmock.Activate()
+
+	httpmock.RegisterResponder(
+		"POST",
+		"http://127.0.0.1:50000/submit/submit",
+		func(req *http.Request) (*http.Response, error) {
+			log.Info("POST", "url", req.URL, "body", req.Body)
+			return httpmock.NewJsonResponse(200, nil)
+		},
+	)
+
+	httpmock.RegisterResponder(
+		"GET",
+		`=~http://127.0.0.1:50000/availability/header/(\d+)`,
+		func(req *http.Request) (*http.Response, error) {
+			log.Info("GET", "url", req.URL)
+			block := uint64(httpmock.MustGetSubmatchAsUint(req, 1))
+			// todo
```

Comment:
These `todo`s are here just because we don't realize the validation of espresso. After we realize that, we need to make the proof or header correctly or the test will fail.

Reply:
Now `todo`s are removed and a comment is left.